### DISCUSSION TITLE
Connecting to Create 3 via Bluetooth in Spyder

### DISCUSSION THREAD

**Discussion by elepowsky** [iRobotEducation/create3_docs#413] [Questions]
### How are you connecting to your Create 3?

Bluetooth (Python SDK)

### Computer(s) Model(s) and Operating System(s)

Dell XPS 13 9310, Windows 10

### Which version of ROS 2 is installed on your computer?

None

### Which firmware version is installed on your robot?

G.4.5

### Which RMW is your robot running?

I don't know

### Are there multiple Create 3 robots connected to your network?

No

### Is multicast enabled?

No

### What is the Adapter Board's USB/BLE Toggle currently switched to?

Bluetooth (default)

### Describe your question.

The Create 3 connects to the Web Playground without any issues and is able to run the examples. Connecting from the command prompt also works without issue. However when trying to run the Python SDK scripts using Spyder, the Create 3 fails to connect.

A simple test script using Bleak in Spyder, with the addition of  `nest_asyncio.apply()`, is able to identify that the Create 3 is available and connect to it. The Create 3 successfully plays the connected/disconnected sound. The test script is below:

```
import asyncio
import nest_asyncio

nest_asyncio.apply()

from bleak import BleakScanner, BleakClient

address = []

async def main():
    global address
    devices = await BleakScanner.discover()
    for d in devices:
        print(d)
        if d.name == 'iRobotCreate3':
            address = d.address
    async with BleakClient(address) as client:
        print('Connected!')

asyncio.run(main())
```

Adding sections to one of the examples (for instance, `get_position.py`) seems to show that the code gets stuck at the `robot = Create3(Bluetooth())` line. If I hardcode the device name and address (from the previous Bleak scan), the code will progress past this line, and then still hangs forever in the next block of code.

Any ideas what could be causing this issue and how to resolve it? Is the Python SDK not compatible with Spyder?

**Comment by miniBloq** [comment]
Hi, by Spyder you are referring to the Spyder IDE (https://www.spyder-ide.org/)?

**Reply by elepowsky** [reply]
Yes, exactly. I'm assuming it's a compatibility issue between asyncio and IPython, but hoping there might be a known workaround specific to interfacing with the Create 3. The alternative, of course, will be to stick with only running the Python scripts from the command prompt.

**Comment by miniBloq** [comment]
That is possible, certainly IPython may present some critical differences that can prevent the BLE backend or the asyncio system from working properly. The desktop Python SDK is a Beta, and it is aimed mainly to the command line. The idea is also to provide a codebase from which users can learn, contribute, and test their own Python code. Sadly, I do not think that we can provide much help with specific third party IDEs like Spyder. IPython is a very nice tool, but it is out of our scope for now, even when we would really like to see it working with the beta SDK.
